### PR TITLE
Support of external members [19704]

### PR DESCRIPTION
**Pull request by richiware**

### CODE REVIEW

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/315/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Mac/229/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/314/)

**Comment by JLBuenoLopez-eProsima** [comment]
[Manual CI run](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual/314/): SUCCESS

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
A few typos in the comments. Implementation LGTM!

File: test/xcdr/external.cpp [hunk 1]
```diff
@@ -0,0 +1,1385 @@
...
+    {
+        //{ Prepare buffer
+        auto buffer =
+                std::unique_ptr<char, void (*)(
+            void*)>{reinterpret_cast<char*>(calloc(20, sizeof(char))), free};
+        FastBuffer fast_buffer(buffer.get(), 20);
+        Cdr cdr(fast_buffer, endianness, get_version_from_algorithm(encoding));
+        //}
+
+        //{ Encode optional not present.
```
[the remark this answers is not in the document]
Comment (richiware) [reply test/xcdr/external.cpp]: Done in 0f9a604

File: test/xcdr/external.cpp [hunk 2]
```diff
@@ -0,0 +1,1385 @@
...
+    //{ Prepare buffer
+    uint8_t tested_stream = 0 + encoding + endianness;
+    auto buffer =
+            std::unique_ptr<char, void (*)(
+        void*)>{reinterpret_cast<char*>(calloc(expected_streams[tested_stream].size(), sizeof(char))), free};
+    FastBuffer fast_buffer(buffer.get(), expected_streams[tested_stream].size());
+    Cdr cdr(fast_buffer, endianness, get_version_from_algorithm(encoding));
+    //}
+
+    //{ Encode optional not present.
```
[the remark this answers is not in the document]
Comment (richiware) [reply test/xcdr/external.cpp]: Done in 0f9a604

File: test/xcdr/external.cpp [hunk 3]
```diff
@@ -0,0 +1,1385 @@
...
+    _locked_ext->text("test3");
+    ASSERT_EQ((*not_empty_ext3).text(), "test2");
+    ASSERT_FALSE(not_empty_ext3.is_locked());
+
+    // Calling assignment over a locked external.
+    external<ExternalClass> locked_ext {new ExternalClass("test"), true};
+    EXPECT_THROW(locked_ext = _locked_ext, exception::LockedExternalAccessException);
+}
+
+//! @test Tests the deference object functions.
```
[the remark this answers is not in the document]
Comment (richiware) [reply test/xcdr/external.cpp]: Done in 0f9a604

File: test/xcdr/external.cpp [hunk 4]
```diff
@@ -0,0 +1,1385 @@
...
+    not_empty_ext2 = _ext;
+
+    ASSERT_TRUE(not_empty_ext2);
+    ASSERT_TRUE(nullptr != not_empty_ext2.get());
+    ASSERT_EQ((*not_empty_ext2).text(), "test2");
+    _ext->text("test3");
+    ASSERT_EQ((*not_empty_ext2).text(), "test3");
+    ASSERT_FALSE(not_empty_ext2.is_locked());
+
+    // Calling assignment over an empty external with an locked external;
```
[the remark this answers is not in the document]
Comment (richiware) [reply test/xcdr/external.cpp]: Done in 0f9a604

File: include/fastcdr/xcdr/external.hpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        return pointer_.get();
+    }
+
+    //! Get shared pointer.
+    std::shared_ptr<T> get_shared_ptr() noexcept
+    {
+        return pointer_;
+    }
+
+    //! Deference object member.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastcdr/xcdr/external.hpp]: Done in 0f9a604

File: include/fastcdr/xcdr/external.hpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            pointer_ = other.pointer_;
+        }
+
+        return *this;
+    }
+
+    //! Destructor
+    ~external() = default;
+
+    //! Deference object.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastcdr/xcdr/external.hpp]: Done in 0f9a604

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/323/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Mac/236/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/322/)

**Comment by richiware** [comment]
Relaunched Mac job success: [link](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Mac/237/)

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Pull request merged**